testbed/tests/app/test_desktop.py

```diff
@@ -370,26 +368,47 @@ async def test_current_window(app, app_probe, main_window):
     window2.content = toga.Box(style=Pack(background_color=CORNFLOWERBLUE))
     window3.content = toga.Box(style=Pack(background_color=FIREBRICK))
 
-    # We don't need to probe anything window specific; we just need
-    # a window probe to enforce appropriate delays.
-    window1_probe = window_probe(app, window1)
-
     window1.show()
     window2.show()
     window3.show()
 
-    await window1_probe.wait_for_window("Extra windows added")
+    await main_window_probe.wait_for_window("Extra windows added")
 
+    # When a window without any dialog is made the current_window,
+    # then `app.current_window` should return the specified window.
     app.current_window = window2
-    await window1_probe.wait_for_window("Window 2 is current")
+    await main_window_probe.wait_for_window("Window 2 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window2
 
     app.current_window = window3
-    await window1_probe.wait_for_window("Window 3 is current")
+    await main_window_probe.wait_for_window("Window 3 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window3
 
+    # When a dialog is in focus, `app.current_window` should
+    # return the window from which the dialog was initiated.
+    info_dialog = toga.InfoDialog("Info", "Some info")
+    app_probe.setup_info_dialog_result(info_dialog)
+
+    app.current_window = window1
+    await main_window_probe.wait_for_window("Window 1 is current")
+    dialog_task = app.loop.create_task(window1.dialog(info_dialog))
+    await main_window_probe.wait_for_window("Displayed window1 modal info dialog")
+    # The public API should report that current window is the specified window
+    if app_probe.supports_current_window_assignment:
+        assert app.current_window == window1
+    # But, the backend should be reporting that the current window is the dialog
+    if toga.platform.current_platform == "macOS":
+        assert "_NSAlertPanel" in str(
+            app._impl.native.keyWindow
+        ), "The dialog is not in focus"
```

Comment:
This test is only to test that the test is actually testing the situation of "in presence of dialog" and not testing the situation where the dialog immediately exits.